Improve Shared Worker Examples (#42)
privacycg/saa-non-cookie-storage: shared-workers.md

## [Maintaining a Session](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support/issues/124)

A website, chat.example, offers a way to maintain active chat session between any sites the embed it using SharedWorkers.

> Before storage partitioning this was possible, but after storage partitioning the when chat.example is embedded on different sites and instanciates a SharedWorker it would no longer be shared.

Reviewer: ```suggestion
Before storage partitioning this was possible, but after storage partitioning the when chat.example is embedded on different sites and instantiates a SharedWorker it would no longer be shared.
```

> Before storage partitioning this was possible, but after storage partitioning the when chat.example is embedded on different sites and instanciates a SharedWorker it would no longer be shared.

Reviewer 2: ```suggestion
Before storage partitioning this was possible, but after storage partitioning when chat.example is embedded on different sites and instanciates a SharedWorker it would no longer be shared.
```

> By prompting the user for permission via `document.requestStorageAccess({SharedWorker: true})` and then instanciating the SharedWorker via the returned handle the worker can be shared across partitioned third-party contexts.

Reviewer 3: This is all accurate technically, but what's the end user's experience? What are they actually trying to do here; what broke for them when partitioning happened; what workarounds could the website deploy without this feature; and what do you want the experience to look like after this feature?

Reviewer: I have a different question: if you have been granted storage access permission, why aren't all shared workers created in the unpartitioned space by default?

Why would you not instead have an option to ask for the shared worker to be partitioned, just like with CHIPS?  The handle design is more complex to use, by far.

Reviewer 2: @martinthomson this is downstream from the decision to partition third-party storage vs. block it by default. Turning default partitioned storage into unpartitioned storage at runtime creates a lot of difficult questions, like what if the developer still has access to e.g. handles for an open IndexedDB database? What if a script calling SAA races with a script creating workers? Having a clean interface with both modes accessible makes things much easier for developers and browsers.

Yes, if we had blocked storage by default, we could have a mechanism to ask for partitioned storage instead, and then we could also do this thing where we open up access to unpartitioned storage in the default namespace, but we have browser consensus on default partitioning instead, and that's just how things went.

Reviewer 3: +1 to @johannhof: the TAG endorsed this overall method of handling non-cookie storage in https://github.com/w3ctag/design-reviews/issues/906#issuecomment-1882664982 (just before @martinthomson joined). It'd be possible to reopen that, but I don't think this PR is the right place.

> By prompting the user for permission via `document.requestStorageAccess({SharedWorker: true})` and then instanciating the SharedWorker via the returned handle the worker can be shared across partitioned third-party contexts.

Reviewer: ```suggestion
By prompting the user for permission via `document.requestStorageAccess({SharedWorker: true})` and then instantiating the SharedWorker via the returned handle the worker can be shared across partitioned third-party contexts.
```

## [Transfering an ArrayBuffer](https://groups.google.com/a/chromium.org/g/blink-dev/c/inRN8tI49O0/m/Q_TE0cw4AAAJ)

A website, worker.example, wants to track and resume work in a first party context using SharedWorkers. `window.postMessage(...)` won't work as it would require inefficient cloning of significant data.

> Before storage partitioning this was possible, but after storage partitioning the when worker.example is embedded in a third-party context and instanciates a SharedWorker it would no longer be shared with workers instanciated in a first-party context.

Reviewer 2: ```suggestion
Before storage partitioning this was possible, but after storage partitioning when worker.example is embedded in a third-party context and instantiates a SharedWorker it would no longer be shared with workers instantiated in a first-party context.
```

> By prompting the user for permission via `document.requestStorageAccess({SharedWorker: true})` and then instanciating the SharedWorker via the returned handle the worker can be shared from a partitioned third-party context to a first-party context if the first-party worker is instanciated with the `sameSiteCookies: 'none'` option.

Reviewer 2: ```suggestion
By prompting the user for permission via `document.requestStorageAccess({SharedWorker: true})` and then instantiating the SharedWorker via the returned handle the worker can be shared from a partitioned third-party context to a first-party context if the first-party worker is instantiated with the `sameSiteCookies: 'none'` option.
```

## [Maintaining a Session](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support/issues/124)

> A website, chat.example, offers a way to maintain active chat session between any sites the embed it using SharedWorkers.

Reviewer 3: I'd start with something like, "Some websites allow users to chat with the site's customer service. In some cases, pieces of a user's natural journey while interacting with the customer service agent lead them to different but associated sites. For example, [put in a real world site so reviewers can actually click through the experience]. This could be implemented by putting a session ID into a cookie, and that would continue to work while cookies are partitioned if the user grants the requestStorageAccess prompt. However, it's more natural and common for sites to use a SharedWorker to manage the chat session. This proposal allows that to work after the same use permission grant."

I actually really dislike the idea of granting storage access to chat.example, since it could then track all across the Web instead of just across the set of sites that naturally share chat sessions. We can't prevent it technically, but maybe the example could use chat.site-a.com instead?

## [Maintaining a Session](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support/issues/124)

Some websites allow users to chat with the site's customer service.
For example, the US internet provider Comcast splits parts of its services across business.comcast.com, xfinity.com, and more.
In some cases, pieces of a user's natural journey while interacting with the customer service agent lead them to different but associated sites.
This could be implemented by putting a session ID into a cookie, and that would continue to work while cookies are partitioned if the user grants storage access.
However, it's more natural and common for sites to use a SharedWorker to manage the chat session.
This proposal allows that to work after the same use permission grant.

A website, company-chat.example, offers a way to maintain active chat session between any sites the embed it using SharedWorkers.
Before storage partitioning this was possible, but after storage partitioning when company-chat.example is embedded on different sites (company-site1.example or company-site2.example) and instantiates a SharedWorker it would no longer be shared.

> By prompting the user for permission via `document.requestStorageAccess({SharedWorker: true})` and then instantiating the SharedWorker via the returned handle the worker can be shared across partitioned third-party contexts.

Reviewer 3: Thanks! I think the introduction matches https://w3ctag.github.io/explainer-explainer/#end-user-need now.

The second paragraph could flow better though: you've already introduced the idea of an active chat session, so you could start with something like, "To implement this, the company embeds a helper site, e.g. `company-chat.example`, into all of its branded sites, and `company-chat.example` uses a `SharedWorker` to communicate between its multiple embedded instances."

## [Transfering an ArrayBuffer](https://groups.google.com/a/chromium.org/g/blink-dev/c/inRN8tI49O0/m/Q_TE0cw4AAAJ)

> Similar to the above case, but perhapse the customer service chat has a video option that can start embedded in the page but could also be popped out into its own window.

Reviewer 2: ```suggestion
Similar to the above case, but perhaps the customer service chat has a video option that can start embedded in the page but could also be popped out into its own window.
```

This could be implemented by putting a session ID into the query params of the opened window.

> However, to allow a more natural transition without audio/video glitches it would be easier to have a SharedWorker seemlessly transition the call over.

Reviewer 2: ```suggestion
However, to allow a more natural transition without audio/video glitches it would be easier to have a SharedWorker seamlessly transition the call over.
```

This proposal allows that to work after a storage access permission grant.

> A website, company-videochat.example, wants to track and resume a video call in a first party context using SharedWorkers. `window.postMessage(...)` won't work as it would require inefficient cloning of significant data.

Reviewer 3: I think I'm not familiar enough with the way transferring ArrayBuffers works: do they transfer between a site and its own `SharedWorker` but not across cross-site iframe boundaries? I guess that makes sense, since different sites are likely to be running in different processes. It'd be good to say that explicitly so nobody else stumbles over it.

Author: window.postMessage essentially forces things to be serialized/deserialized as JSON, where SharedWorker can take complex objects like ArrayBuffers intact.

Reviewer 3: That's not intrinsically true, if you use the [`transfer`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#transfer) parameter to detach the source `ArrayBuffer`. I definitely believe there's something getting in the way, but I suspect it's process boundaries, rather than any difference between Windows and Workers.

Author: I had no idea that was possible, cool! The documentation makes it sound like it works to transfer between threads and not processes, yeah.

Author: Okay, so it looks like you can send them cross origin. I suppose the issue is that this does require cloning the data if you need it in both renderers at once.
